### PR TITLE
Add per realm stats

### LINKED ISSUE #410: Realm level stats
### TL;DR

Add charts AND graphs of how many codes a realm has issued per day.

Bonus points for graphing against number of codes that were claimed.

/assign @jeremyfaller 
/kind enhancement

### PR DESCRIPTION
**Pull request by jeremyfaller**
Adds per realm stats, and a simple visualization of the codes issued and
claimed for the past month.

Fixes #410

**Release Note**

```release-note
Adds per-realm stats, and a simple visualization of the last 30 days of issued and claimed codes.
```

### CODE REVIEW

File: cmd/server/assets/header.html
Comment (sethvargo) [inline cmd/server/assets/header.html:188]: We might want to move this above Settings

File: pkg/controller/realmstats/index.go
Comment (sethvargo) [inline pkg/controller/realmstats/index.go:53]: Debug

File: pkg/controller/realmstats/realmstats.go
Comment (sethvargo) [inline pkg/controller/realmstats/realmstats.go:36]:
```suggestion
	logger := logging.FromContext(ctx).Named("realmstats")
```

File: pkg/database/realm.go
Comment (sethvargo) [inline pkg/database/realm.go:614]: Per Go guidelines, I think this should just be `Stats()`?

File: pkg/database/realm_stats.go
Comment (sethvargo) [inline pkg/database/realm_stats.go:23]: Date and RealmID probably need NOT NULL constraints

File: pkg/database/token.go
Comment (sethvargo) [inline pkg/database/token.go:208]: I think we might want to do this inside a callback instead of here.

File: pkg/database/vercode.go
Comment (sethvargo) [inline pkg/database/vercode.go:107]: Yea, if you move this into a callback (see the other stats calculations), we can avoid the duplication.

File: pkg/database/realm.go
Comment (jeremyfaller) [reply pkg/database/realm.go:614]: Done

File: pkg/controller/realmstats/index.go
Comment (mikehelmick) [reply pkg/controller/realmstats/index.go:53]: the printf is still there

File: pkg/database/token.go
Comment (mikehelmick) [reply pkg/database/token.go:208]: agreed - I think it's best to not have stats update required in the transaction

Comment (jeremyfaller) [reply pkg/database/token.go:208]: Done

File: pkg/database/vercode.go
Comment (jeremyfaller) [reply pkg/database/vercode.go:107]: Done.

File: pkg/controller/realmstats/index.go
Comment (jeremyfaller) [reply pkg/controller/realmstats/index.go]: Sorry, I hadn't done a push. My workflow has been to tweak the comment after I update my code. Apparently, messages are sent when I do that. I'll tweak my workflow to not update comments until the new code's been uploaded. Sorry for the spam.

File: pkg/database/realm.go
Comment (sethvargo) [inline pkg/database/realm.go:623]:
```suggestion
		Where("(date >= ? AND date <= ?)", start, stop).
```

I don't trust gorm sometimes...

File: pkg/database/vercode.go
Comment (sethvargo) [reply pkg/database/vercode.go:107]: Did you push? I'm not seeing any changes.

**Comment by jeremyfaller** [comment]
/hold

Hold on this right now. I don't like the error handling I've put in vercode.go. I'm going to add some more testing.

File: pkg/database/vercode.go
Comment (sethvargo) [inline pkg/database/vercode.go:120]: Hmm - doing this in BeforeSave puts us in a spot where we'd increment the stats, but then some validation could prevent the code from actually being claimed.

Comment (jeremyfaller) [reply pkg/database/vercode.go:120]: Ok, I'm going to a side value in VerificationCode. This is too clunky.

Comment (sethvargo) [reply pkg/database/vercode.go:120]: Okay - what are your thoughts on making a `Claim` function on the verification code (or on the database) that handles all this logic?

Comment (jeremyfaller) [reply pkg/database/vercode.go]: LMK what you think of this. It's clunky, but with Hooks (and ORM with dirty) this feels the best we can do. I'm happy to move to a Claim API if you think that's best.

File: pkg/database/vercode.go
Comment (mikehelmick) [inline pkg/database/vercode.go:80]: why not use v.CreatedAt?

File: pkg/database/vercode.go
Comment (mikehelmick) [inline pkg/database/vercode.go:68]: not sure I understand why this is needed in addition to claimed?

Comment (jeremyfaller) [reply pkg/database/vercode.go:68]: Seth had wanted to use Hooks for updating the stats counts, removing the straight DB update someplace else.

If we do that our ORM doesn't know about dirty bits. So, if the VerificationCode has Claimed set, we'll update the stats whenever the hook is called. Now, we could query the DB, and get the old value -- but it's a question of timing. If we do it BeforeSave, we don't know that we're actually going to do the write. If we do it AfterSave, we don't know if Claimed was set BEFORE the update happened. It's a mess.  (This discussion ignores the fact that testing exercises the API in a manner the app never will.)

So, there's two ways out of it -- having this stupid variable, and continuing to use Hooks; or have a Claim function that does this. Hooks are nice because they're auto-updated and you never have to remember to call Claim. The function is nice because it doesn't have this crap floating around.

I am shooting the middle on all approaches here, and as you and Seth have the larger picture, I defer to your judgements (right now!).

File: pkg/database/vercode.go
Comment (jeremyfaller) [reply pkg/database/vercode.go:80]: Oh, better. on it.

File: pkg/database/vercode.go
Comment (mikehelmick) [reply pkg/database/vercode.go:68]: to properly use this as a dirty bit, don't we need to set AreClaiming = false when updating the stats?

otherwise this problem persists

or only set it to true when the stats are updated (actually I like that better, because it enables backfill on stats later if we need to do that)

Comment (jeremyfaller) [reply pkg/database/vercode.go:68]: AreClaiming will default to false (go runtime) for all values read from the DB.

Backfill is a nice side effect I hadn't considered.

Comment (mikehelmick) [reply pkg/database/vercode.go:68]: except, we're saving it to the database

Comment (jeremyfaller) [reply pkg/database/vercode.go:68]: I thought only values listed as "gorm" are saved!!!!

Comment (jeremyfaller) [reply pkg/database/vercode.go:68]: argh

Comment (sethvargo) [reply pkg/database/vercode.go:68]: Hmmmmmm. After seeing that we need to track an additional field, I don't think this is the best approach. I think adding a `verificationcode.Claim()` or `database.ClaimVerificationCode` would be cleaner. Sorry for the turmoil.

**Comment by jeremyfaller** [comment]
/hold

Holding AGAIN. while I work out what I don't understand about gorm.

File: pkg/database/vercode.go
Comment (mikehelmick) [reply pkg/database/vercode.go:68]: To not have it be part of the database model, you have to explicitly tag as ignore

https://gorm.io/docs/models.html#Fields-Tags

File: pkg/database/vercode.go
Comment (jeremyfaller) [reply pkg/database/vercode.go]: Moved to claim. LMK.

File: pkg/database/token.go
Comment (mikehelmick) [inline pkg/database/token.go:191]: this needs to be in tx not db!!!

**Review by mikehelmick — CHANGES_REQUESTED** [review]

File: pkg/database/vercode.go
Comment (sethvargo) [inline pkg/database/vercode.go:112]: We don't currently truncate the other times - should we be?

File: pkg/database/vercode.go
Comment (sethvargo) [inline pkg/database/vercode.go:121]: I think we should drop this log message (and the dependency on `db`) and return the error instead. That'll make it bubble as an internalerror instead of silently swallowing right now.

File: pkg/database/vercode.go
Comment (jeremyfaller) [reply pkg/database/vercode.go:112]: yeah, prolly. With the suggested change here (to use the date from the code), I think all things want truncated dates.

File: pkg/database/vercode.go
Comment (jeremyfaller) [reply pkg/database/vercode.go:121]: you got it.

File: pkg/database/vercode.go
Comment (sethvargo) [inline pkg/database/vercode.go:106]: We should probably guard this with a sanity:

```go
if v.RealmID != 0 {
```

can be a follow-up

File: pkg/database/vercode.go
Comment (sethvargo) [inline pkg/database/vercode.go:122]: It'd be nice to wrap this error like `fmt.Errorf("failed to claim code: %w")`

Comment (sethvargo) [reply pkg/database/vercode.go:122]: Can be a follow-up

File: pkg/database/vercode.go
Comment (sethvargo) [inline pkg/database/vercode.go:120]: This API is a little weird to me now, and I'm not sure the comment matches. While this sets `Claimed: true` on the Verification Code, it does not _save_ that code in the database. The caller still needs to call `db.Save(v)`. That feels subtle and could lead to some bugs. Furthermore, we're updating the realm stats _before_ successfully marking the code as claimed. I'm thinking this flow should be more like:

1. Expiration checks
1. Claimed check
1. Mark as claimed on struct
1. Save struct to database (stop if error)
1. Update stats
1. Return

Alternatively, we need to document this better. What do you think?

**Pull request closed without merging**
